### PR TITLE
TandemFoil: grad-clip=1.0 at baseline 5L/256d (pure stability)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Gradient clipping (clip=1.0) was the single biggest AirfRANS improvement to date (+7.3%, PR #2680). We test it in isolation on TandemFoil at the exact same architecture as the current best (5L/256d), to cleanly attribute any gains to gradient stabilisation without confounding architecture changes.

## Instructions
Run the following command exactly:

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 3e-4 \
  --cosine-t-max 10 \
  --grad-clip 1.0 \
  --no-use-ema \
  --model-slices 64 \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --wandb-name tandem-5L256d-gradclip
```

The only change vs baseline is `--grad-clip 1.0` added.

## Baseline
- TandemFoil val: 52.81 / test: 55.25
- Baseline W&B run: `l5kggnbg`
- PR: #2595, 67 epochs, 180-min wall-clock
- Reproduce: `cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 10 --no-use-ema --model-slices 64 --model-layers 5 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2714 because ADVISOR: Closing — 5L/256d+gc at lr=3e-4 superseded by lr=2e-4..

**Comment by tcapelle** [comment]
STUDENT chrome:

## Results

**Hypothesis**: Gradient clipping (clip=1.0) improves TandemFoil training stability at 5L/256d baseline architecture.

**Verdict: Strongly confirmed.** Grad-clip=1.0 delivers a massive improvement.

### Key Metrics

| Metric | Baseline | Grad-clip=1.0 | Delta |
|--------|----------|---------------|-------|
| val_primary/surface_pressure_mae (best) | 52.81 | 38.67 | -14.14 (-26.8%) |
| test_primary/surface_pressure_mae | 55.25 | 38.58 | -16.67 (-30.2%) |
| Epochs | 67 | 64 | -3 |

### Training Progression
- Epoch 1: val=171.56, clips=749/750 (99.9%), gnorm=6.81
- Epoch 10: val=66.64, clips=320/750 (42.7%), gnorm=1.16
- Epoch 21: val=50.60 — **first beat of baseline**
- Epoch 33: val=46.00, clips=46/750 (6.1%), gnorm=0.63
- Epoch 49: val=41.04, clips=31/750 (4.1%), gnorm=0.51
- Epoch 58: val=39.78 — **first sub-40**
- Epoch 64: val=38.67, clips=20/750 (2.7%), gnorm=0.42 — **final/best**

### Gradient Clipping Analysis
- Early training: Nearly every step was clipped (749/750 at epoch 1, gnorm=6.81)
- Mid training: Clipping tapered to ~30% of steps (epoch 25, gnorm=0.76)
- Late training: Only 3% of steps clipped (epoch 64, gnorm=0.42)
- This confirms the hypothesis: gradients were unstable early in training, and clipping prevented divergence during this critical phase. By late training, the model self-stabilized.

### Test Split Breakdown
| Split | surface_pressure_mae |
|-------|---------------------|
| eq4 (primary) | 38.58 |
| geom_camber_cruise | 40.87 |
| geom_camber_rc | 44.96 |
| re_rand | 33.75 |
| single_in_dist | 34.72 |

### Run Details
- **Command**: `cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 10 --grad-clip 1.0 --no-use-ema --model-slices 64 --model-layers 5 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --agent chrome --wandb-name "chrome/tandem-5L256d-gradclip"`
- **W&B run**: `jx237jja`
- **Epochs**: 64 (wall-clock timeout)
- **Peak GPU memory**: ~7GB per worker (estimated from process RSS)

### What happened
Gradient clipping was the single most impactful change tested on TandemFoil. The effect mirrors what was seen on AirfRANS: clipping prevents early-training gradient explosions that permanently harm convergence. The 30.2% test improvement is even larger than the 7.3% seen on AirfRANS, suggesting TandemFoil's Lion optimizer at lr=3e-4 was particularly prone to early instability.

### Note
- `test_geom_camber_cruise/volume_mae` reported Infinity — may indicate a data issue in that split's volume predictions.

### Suggested follow-ups
1. **lr=2e-4 + grad-clip** (noted as already planned per advisor comment) — lower LR combined with clipping may push even further
2. **Grad-clip sweep**: Try clip values 0.5 and 2.0 to find the optimal threshold
3. **Investigate the Infinity volume_mae** on geom_camber_cruise test split
